File: Android/app/src/main/java/com/nfsindustries/cryptocurrencymonitor/service/bitcoin/CoinmarketCapAPI.java
package com.nfsindustries.cryptocurrencymonitor.service.bitcoin;

import android.content.Context;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.nfsindustries.cryptocurrencymonitor.CryptoCurrencyMonitor;
import com.nfsindustries.cryptocurrencymonitor.deserializer.CoinmarketcapDeserializer;
import com.nfsindustries.cryptocurrencymonitor.model.CryptoCurrencyModel;
import com.nfsindustries.cryptocurrencymonitor.utils.CachingControlInterceptor;
import com.nfsindustries.cryptocurrencymonitor.utils.Constants;

import java.io.File;

import okhttp3.Cache;
import okhttp3.OkHttpClient;
import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.GET;
import retrofit2.http.Path;

/**
 * Created by maxmo on 23/5/2017.
 */

public interface CoinmarketCapAPI {

    Context context = CryptoCurrencyMonitor.getAppContext();
    long SIZE_OF_CACHE = 10 * 1024 * 1024; // 10 MiB
    Cache cache = new Cache(new File(context.getCacheDir(), "http"), SIZE_OF_CACHE);
    CachingControlInterceptor networkInterceptor = new CachingControlInterceptor();

    OkHttpClient client = new OkHttpClient.Builder().cache(cache).addNetworkInterceptor(networkInterceptor).build();

    Gson gson = new GsonBuilder()
            .setLenient()
            .registerTypeAdapter(CryptoCurrencyModel.class, new CoinmarketcapDeserializer())
            .create();

    Retrofit retrofit = new Retrofit.Builder()
            .baseUrl(Constants.COIN_MARKETCAP_BASE_URL)
            .client(client)
            .addConverterFactory(GsonConverterFactory.create(gson))
            .build();

    @GET("{cryptocurrency}?convert=USD")
    Call<CryptoCurrencyModel> getCurrentIndex(@Path("cryptocurrency") String cryptocurrency);
}

File: Android-Kotlin/app/src/main/java/com/nfsindustries/cryptocurrencymonitor/service/coinmarket/CoinmarketCapAPI.kt
package com.nfsindustries.cryptocurrencymonitor.service.coinmarket

import com.google.gson.GsonBuilder
import com.nfsindustries.cryptocurrencymonitor.utils.Constants
import com.nfsindustries.cryptocurrencymonitor.CryptoCurrencyMonitor
import com.nfsindustries.cryptocurrencymonitor.deserializer.CoinmarketcapDeserializer
import com.nfsindustries.cryptocurrencymonitor.model.CryptoCurrencyModel
import com.nfsindustries.cryptocurrencymonitor.utils.CachingControlInterceptor

import java.io.File

import okhttp3.Cache
import okhttp3.OkHttpClient
import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.GET
import retrofit2.http.Path

/**
 * Interface to get Crypto Currency data from CoinMarket API.
 * Created by MaxMouse on 9/21/17.
 */
interface CoinmarketCapAPI {

    @GET("{cryptocurrency}?convert=USD")
    fun getCurrentIndex(@Path("cryptocurrency") cryptocurrency: String): Call<CryptoCurrencyModel>

    companion object {

        val CACHE_SIZE = (20 * 1024 * 1024).toLong() // 20 MiB
        val COIN_MARKETCAP_BASE_URL = "https://api.coinmarketcap.com/v1/"
        val TICKER = "ticker"
        val context = CryptoCurrencyMonitor.appContext
        val cache = Cache(File(context?.getCacheDir(), "http"), CACHE_SIZE)
        val networkInterceptor = CachingControlInterceptor()

        //TODO FIX this instant run bullshit

        val client = OkHttpClient.Builder().cache(cache).addNetworkInterceptor(networkInterceptor).build()

        val gson = GsonBuilder()
                .setLenient()
                .registerTypeAdapter(CryptoCurrencyModel::class.java, CoinmarketcapDeserializer())
                .create()

        val retrofit = Retrofit.Builder()
                .baseUrl(COIN_MARKETCAP_BASE_URL + TICKER + "/")
                .client(client)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build()
    }
}